docs/contributing/contributing.md

```diff
@@ -108,79 +108,73 @@ $ cabal run haskell-language-server:func-test -- -p "hlint enables"
 
 ## Using HLS on HLS code
 
-Project source code should load without `hie.yaml` setup.
+Refer to the [HLS project configuration guidelines](../configuration.md#configuring-your-project-build) as they also apply to the HLS project itself.
 
-In other cases:
-
-1. Check if `hie.yaml` (& `hie.yml`) files left from previous configurations.
-
-2. If the main project needs special configuration, note that other internal subprojects probably also would need configuration.
-
-To create an explicit configuration for all projects - use [implicit-hie](https://github.com/Avi-D-coder/implicit-hie) generator directly:
-
-```shell
-gen-hie > hie.yaml  # into the main HLS directory
-```
-
-that configuration should help.
-
-3. Inspect & tune configuration explicitly.
-
-[Configuring project build](../configuration.md#configuring-your-project-build) applies to HLS project source code loading just as to any other.
-
-Note: HLS may implicitly detect codebase as a Stack project (see [hie-bios implicit configuration documentation](https://github.com/haskell/hie-bios/blob/master/README.md#implicit-configuration)). To use Cabal, try creating an `hie.yaml` file:
+Note: HLS implicitly detects the HLS codebase as a Stack project (since there is a `stack.yaml` file).
+If you want HLS to use Cabal, create this `hie.yaml` file at the root of the project:
 
 ```yaml
 cradle:
   cabal:
 ```
 
-### Manually testing your hacked HLS
-If you want to test HLS while hacking on it, follow the steps below.
-
-#### Using Cabal
+## Manually testing your hacked HLS
+If you want to test HLS while hacking on it (you can even test it on HLS codebase itself, see previous section), you need to:
 
-- Whenever you want to build HLS, call `cabal install exe:haskell-language-server --overwrite-policy=always`. 
-  At the end of the output you will find the path the HLS executable was installed to, i.e.:
+1. (Once) Find the path to the hacked HLS you build
+2. (Once) Configure your editor to use it
+3. (Every time you change the HLS code) Rebuild HLS
+4. (Every time you change the HLS code) Restart the LSP workspace
 
-  ```
-  ...
-  Resolving dependencies...
-  Symlinking 'haskell-language-server' to
-  '/home/user/.cabal/bin/haskell-language-server'
-  Symlinking 'haskell-language-server-wrapper' to
-  '/home/user/.cabal/bin/haskell-language-server-wrapper'
-  ```
+### Find the path to the hacked HLS you build
+Note that unless you change the GHC version or the HLS version between builds, the path should remain the same, this is why you need to set it only once.
 
-  In this example output, the path would be `/home/user/.cabal/bin/haskell-language-server`.
+#### Using Cabal
+Run:
+```shell
+$ cabal build exe:haskell-language-server && cabal list-bin exe:haskell-language-server
+[..]
+<some long path>/haskell-language-server
+```
```

Comment:
I know I am a little bit late to the party, but I feel like using the `cabal build` binary for testing can be slightly unexpected.
When you switch branches and run `cabal build` just to make sure it compiles, or to speed up HLS to load everything, then you suddenly update your test binary.
Such behaviour might be unexpected to non-expert users, perhaps even expert users. I, for one, definitely prefer `cabal install` since I have a couple of local HLS installations lying around.

That's why we previously recommended using `cabal install`, updating your tested HLS binary is a very conscious decision. Additionally, it allows you to set program-suffixes and other custom locations.

Reply:
Arguably the opposite argument is as inconvenient for others 😅
1. `cabal install` will install in a global location outside of the project, so it makes it harder to keep things isolated
2. It might replace the stable binary they had installed in the past, and they might get confused about how to get back to the stable binary
3. With the `cabal build` approach, you can "undo" all the steps in the documentation and land back on your original setup. With the `cabal install` approach, "undoing" the steps may not be enough to get back to the original setup (because you also need to know how to get back the original binary, which you may have installed long ago, and you did not know which version you had).

I guess different people have different priorities 😅 And it's completely ok by the way!
In my opinion the new version of the documentation is more welcoming towards non-experts (ability to undo), and while some experts may get surprised by the scenario you mentioned, I think they would be able to identify and work around.

In any case it might also be beneficial to add this caveat to the documentation, if you suspect many will get confused by it

Reply:
Fair enough, the arguments are also convincing. Ideally, we'd have a command `cabal build --copy-to-local-bindir`, that copies the binary to `dist-newstyle/bin/` or something similar. 

What do you think about instructions such as `cabal install --installdir dist-newstyle/bin/`? Arguably, we shouldn't touch `dist-newstyle`, but I think that's relatively safe.

Reply:
FWIW I use `cabal build` and it's fine, but it seems reasonable to include both as alternatives?